test/java/seedu/addressbook/data/AddressBookTest.java

```diff
@@ -0,0 +1,275 @@
+        UniqueTagList tagsInPersons = getAllTags(persons);
+        UniqueTagList tags = addressBook.getAllTags();
+        
+        // all person tags are equal to some tag in the master list
+        boolean allPersonTagsValid = true;
+        for (Tag personTag : tagsInPersons) {
+            // particular person tag is equal to some tag in the master list
+            boolean personTagValid = false;
+            for (Tag tag : tags) {
+                personTagValid = personTagValid || (personTag == tag);
+            }
+            allPersonTagsValid = allPersonTagsValid && personTagValid;
+        }
+        
+        assertTrue(allPersonTagsValid);
+    }
+
+    @Test
+    public void addPerson_personNotInList_addsNormally() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+        
+        // add person whose tag is in tag list
+        assertFalse(addressBook.containsPerson(charlieDouglas));
+        addressBook.addPerson(charlieDouglas);
+        assertTrue(addressBook.containsPerson(charlieDouglas));
+        
+        // add person whose tag is not in tag list
+        assertFalse(addressBook.containsPerson(davidElliot));
+        addressBook.addPerson(davidElliot);
+        assertTrue(addressBook.containsPerson(davidElliot));
+    }
+
+    @Test
+    public void addPerson_personAlreadyInList_throwsDuplicatePersonException() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+        
+        thrown.expect(DuplicatePersonException.class);
+        addressBook.addPerson(aliceBetsy);
+    }
+
+    @Test
+    public void addTag_tagNotInList_addsNormally() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+
+        assertFalse(addressBook.containsTag(tagEconomist));
+        addressBook.addTag(tagEconomist);
+        assertTrue(addressBook.containsTag(tagEconomist));
+    }
+
+    @Test
+    public void addTag_tagAlreadyInList_throwsDuplicateTagException() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+        
+        thrown.expect(DuplicateTagException.class);
+        addressBook.addTag(tagMathematician);
+    }
+
+    @Test
+    public void containsPerson() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+
+        assertTrue(addressBook.containsPerson(aliceBetsy));
+        assertFalse(addressBook.containsPerson(charlieDouglas));
+    }
+
+    @Test
+    public void containsTag() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+        
+        assertTrue(addressBook.containsTag(tagScientist));
+        assertFalse(addressBook.containsTag(tagEconomist));
+    }
+
+    @Test
+    public void removePerson_personExists_removesNormally() throws Exception {
+        AddressBook addressBook = createDefaultAddressBook();
+
+        assertTrue(addressBook.containsPerson(aliceBetsy));
+        addressBook.removePerson(aliceBetsy);
```

Comment:
Should at least check the size to ensure no more than the target person was removed.